tools/bpf/bpftool/gen.c

```diff
@@ -1097,9 +1100,887 @@ static int do_help(int argc, char **argv)
+				hashmap__for_each_entry(reloc_type->members, member_entry, bkt2) {
+					struct btf_reloc_member *reloc_member;
+					struct btf_member *btf_member;
+
+					reloc_member = member_entry->value;
+					btf_member = btf_members(btf_type) + reloc_member->idx;
+
+					memcpy(btf_members(btf_type_cpy) + index, btf_member,
+					       sizeof(struct btf_member));
+
+					index++;
+				}
+			}
+
+			/* set new vlen */
+			btf_type_cpy->info = btf_type_info(btf_kind(btf_type_cpy), nmembers,
+							   btf_kflag(btf_type_cpy));
+
+			err = btf__add_type(btf_new, info->src_btf, btf_type_cpy);
+			free(btf_type_cpy);
+		} else {
+			err = btf__add_type(btf_new, info->src_btf, btf_type);
+		}
+
+		if (err < 0)
+			goto out;
+
+		new_id = err;
+
+		/* add ID mapping */
+		err = btf_reloc_id_add(info, reloc_type->id, new_id);
+		if (err)
+			goto out;
+	}
+
+	/* second pass: fix up type ids */
+	for (unsigned int i = 0; i < btf__type_cnt(btf_new); i++) {
+		struct btf_member *btf_member;
+		struct btf_type *btf_type;
+		struct btf_param *params;
+		struct btf_array *array;
+
+		btf_type = (struct btf_type *) btf__type_by_id(btf_new, i);
+
+		switch (btf_kind(btf_type)) {
+		case BTF_KIND_STRUCT:
+		case BTF_KIND_UNION:
+			for (unsigned short j = 0; j < btf_vlen(btf_type); j++) {
+				btf_member = btf_members(btf_type) + j;
+				btf_member->type = btf_reloc_id_get(info, btf_member->type);
+			}
+			break;
+		case BTF_KIND_PTR:
+		case BTF_KIND_TYPEDEF:
+		case BTF_KIND_VOLATILE:
+		case BTF_KIND_CONST:
+		case BTF_KIND_RESTRICT:
+		case BTF_KIND_FUNC:
+		case BTF_KIND_VAR:
+			btf_type->type = btf_reloc_id_get(info, btf_type->type);
+			break;
+		case BTF_KIND_ARRAY:
+			array = btf_array(btf_type);
+			array->index_type = btf_reloc_id_get(info, array->index_type);
+			array->type = btf_reloc_id_get(info, array->type);
+			break;
+		case BTF_KIND_FUNC_PROTO:
+			btf_type->type = btf_reloc_id_get(info, btf_type->type);
+			params = btf_params(btf_type);
+			for (unsigned short j = 0; j < btf_vlen(btf_type); j++)
+				params[j].type = btf_reloc_id_get(info, params[j].type);
+			break;
+		default:
+			break;
+		}
+	}
+
+	return btf_new;
+
+out:
+	btf__free(btf_new);
+	return ERR_PTR(err);
+}
+
+#define MAX_OBJECTS 128
+
+static int is_file(const char *path)
+{
+	struct stat st;
+
+	if (stat(path, &st) < 0)
+		return -1;
+
+	switch (st.st_mode & S_IFMT) {
+	case S_IFDIR:
+		return 0;
+	case S_IFREG:
+		return 1;
+	default:
+		return -1;
+	}
+
+	return -1;
+}
+
+static int generate_btf(const char *src_btf, const char *dst_btf, const char *objspaths[])
+{
+	struct btf_reloc_info *reloc_info;
+	struct bpf_object *obj;
+	struct btf *btf_new = NULL;
+	int err;
+
+	struct bpf_object_open_opts ops = {
+		.sz = sizeof(ops),
+		.btf_custom_path = src_btf,
+	};
+
+	reloc_info = btfgen_reloc_info_new(src_btf);
+	err = libbpf_get_error(reloc_info);
+	if (err) {
+		p_err("failed to allocate info structure");
+		goto out;
+	}
+
+	for (int i = 0; objspaths[i] != NULL; i++) {
```

Comment:
This leads to a segfault, I passed the length here as an alternative (see the other suggestions).

```suggestion
	for (int i = 0; i < objs_len != NULL; i++) {
```

Reply:
Oh, good finding Lore!

`i < objs_len != NULL` doesn't make much sense tho, but it is clear what you meant 💯

Alternatively (without changing the signature), just check `objspaths` before with an `if (!objspaths)` should prolly do it...

Reply:
Yeah this was supposed to be `i < objs_len` 😄

it can’t be done with `if (!objspaths)` because we need to know if an element exists at a certain index not it the array was initialized.

Reply:
Hmmmm, `if (!objspath)` to clamp, then `objspaths[i] != '\0'` checks?

On the other side, using the length, we should also check that such a length equals the actual one.

Reply:
objspath is an array of strings, not a string

Reply:
Ok, what about this inside the function then?

```c
if (!objspath)
    // error out

int len = sizeof(objspath)/sizeof(objspath[0]);

for(...; i < len; ...)
```

Reply:
Had to move out of the bed, but I wanted to double-check and explain me better :)

```c
#include <stdio.h>
#include <string.h>

#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))

void prova1(char *objspath[])
{
    if (!objspath)
    {
        printf("nullo1\n");
        return;
    }
    int i;
    for (i = 0; objspath[i]; ++i)
    {
        printf("%s\n", objspath[i]);
    }
}

int main()
{
    prova1(NULL);

    char *objspath[] = {"ab", "bc", "cd", 0};
    prova1(objspath);

    return 0;
}
```

Reply:
I think all what is missing is initializing the variable at L1901 -> `const char *objs[MAX_OBJECTS] = {}`, then the NULL check should work fine.

Reply:
Btw, I actually changed it to allocated `objs` using malloc and got rid of the 128 MAX_OBJS limitation.